Add env variable to cmd flags

cyberbono3 opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

Added v.AutomaticEnv() to Viper field of client.Context.
Implemented various test cases to test priority rules `flag > env var > config` for client config.

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

closes: #8179

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

cyberbono3 commented on client/config/cmd_test.go:
```diff
@@ -0,0 +1,61 @@
+
+var ErrWrongNumberOfArgs = fmt.Errorf("wrong number of arguments")
+
+func Test_runConfigCmdTwiceWithShorterNodeValue(t *testing.T) {
```
This is boilerplate for cmd testing. You can find cmd testing [here](https://github.com/cosmos/cosmos-sdk/blob/cyberbono3/8179-add-env-vars/server/util_test.go) as well. What kind of automatic testing do we want to  implment here?

amaurym replied:
Good idea to add os.setEnv tests for clietn config too

amaurym commented on client/config/cmd_test.go:
```diff
@@ -0,0 +1,61 @@
+
+	// Init command config
+	cmd := Cmd()
+	assert.NotNil(t, cmd)
```
prefer require to assert

amaurym commented on client/config/cmd_test.go:
```diff
@@ -0,0 +1,61 @@
+	configHome, cleanup := tmpDir(t)
+	defer cleanup()
+	_ = os.RemoveAll(filepath.Join(configHome, "config"))
+	viper.Set(flags.FlagHome, configHome)
```
here we shou;dn't use global viper

amaurym commented on client/config/cmd_test.go:
```diff
@@ -0,0 +1,61 @@
+	err := cmd.RunE(cmd, []string{"node", "tcp://localhost:26657"})
+	assert.Nil(t, err)
+
+	err = cmd.RunE(cmd, []string{"node", "--get"})
```
Get output value of cmd, check that it's equal to `"tcp://localhost:26657"`

cyberbono3 commented:
@amaurym Could you give your feedback, please?

amaurym commented on client/config/config_test.go:
```diff
@@ -0,0 +1,119 @@
+}
+
+/*
+Second
+env var > config WORKS
+./build/simd config node // tcp://localhost:127 //done already
+NODE=tcp://localhost:1 ./build/simd q staking validators
+Error: post failed: Post "http://localhost:1": dial tcp 127.0.0.1:1: connect: connection refused
+
+Third
+flags > env var > config WORKS
```
I'm a bit confused by these second, third...
if `flags > env var > config WORKS` then this implies `env var > config WORKS`, right?

I would just create on Test function, with multiple test cases. Also, why are we setting env var in `initcontext`? I would think there are some test cases without env var.

amaurym commented on client/config/config_test.go:
```diff
@@ -0,0 +1,104 @@
+		WithHomeDir(home).
+		WithViper()
+
+	clientCtx.Viper.BindEnv(nodeEnv)
```
```suggestion
	// If the NODE env variable is set in test, read it.
	clientCtx.Viper.BindEnv(nodeEnv)
```

amaurym commented on client/config/config_test.go:
```diff
@@ -0,0 +1,104 @@
+			err := cmd.ExecuteContext(ctx)
+
+			require.Error(t, err)
+			require.Contains(t, err.Error(), tc.expNode)
```
could you add a comment on this, on how the `err.Error()` string looks like?

amaurym commented on client/config/cmd_test.go:
```diff
@@ -0,0 +1,61 @@
+package config
```
Up?

amaurym commented on client/config/config_test.go:
```diff
@@ -0,0 +1,104 @@
+	cmd := config.Cmd()
+	// NODE=http://localhost:1 ./build/simd config node http://localhost:2
+	cmd.SetArgs([]string{"node", testNode2})
+	require.NoError(t, cmd.ExecuteContext(ctx))
```
I believe you can use `testutil/cli.ExecTestCLICmd`, which might save a couple of lines

cyberbono3 commented:
in `address comments, r4r` commit I deleted cmd_test.go and modified 'config_test.go'. 
Additionally, I submitted some temporary files (`client/keys/home/keyring-test/746b6a4424a328627f9d10020d53a82765d6642a.address`  etc) by mistake. I have no idea why they have been created in my branch. They have to be omitted.

amaurym reviewed (approved):
Nice!

amaurym reviewed (changes requested):
actually, before auto-merging, let me just test manually on the CLI too

amaurym commented on client/context.go:
```diff
@@ -229,6 +229,7 @@ func (ctx Context) WithInterfaceRegistry(interfaceRegistry codectypes.InterfaceR
 // client-side config from the config file.
 func (ctx Context) WithViper() Context {
 	v := viper.New()
+	v.AutomaticEnv()
```
Also, I'm reading https://github.com/cosmos/cosmos-sdk/issues/8179#issuecomment-747087678. We should actually also allow prefixes, e.g. `GA_CHAIN_ID` or `AKASH_NODE`. What i propose with that WithViper takes a `prefix string` argument, and then we `v.SetEnvPrefix(prefix)`

amaurym commented on simapp/simd/cmd/root.go:
```diff
@@ -47,7 +47,7 @@ func NewRootCmd() (*cobra.Command, params.EncodingConfig) {
 		WithBroadcastMode(flags.BroadcastBlock).
 		WithHomeDir(simapp.DefaultNodeHome).
-		WithViper()
+		WithViper("")
```
```suggestion
		WithViper("") // In simapp, we don't use any prefix for env variables.
```

amaurym reviewed (approved):
tACK.

i tested:
```
# setting manually WithViper("ABC")
ABC_NODE=localhost:1 simd q staking validators
```

amaurym commented:
@cyberbono3 Could you fix the lint errors?

cyberbono3 commented:
> @cyberbono3 Could you fix the lint errors?

Sure. I have fixed some of them. 

Issues:
1. I have no idea why [TestPaginatedVotesQuery](https://github.com/cosmos/cosmos-sdk/pull/9040/checks?check_run_id=2327967052) fails. 
2. This no file shown [here](https://github.com/cosmos/cosmos-sdk/pull/9040/checks?check_run_id=2327963639) that has to gofmted.

amaurym commented:
> I have no idea why TestPaginatedVotesQuery fails.

It's flaky, see #9010 

> This no file shown here that has to be gofmted.

See https://github.com/cosmos/cosmos-sdk/pull/9040/files#diff-146f73b0327c426deb3126e44f9234c7a6fe190c1bcb4c98ab8366b3485922edR116

amaurym commented on client/config/config_test.go:
```diff
@@ -0,0 +1,107 @@
+
+	clientCtx.Viper.BindEnv(nodeEnv)
+	if envVar != "" {
+		os.Setenv(nodeEnv, testNode1)
```
```suggestion
		os.Setenv(nodeEnv, envVar)
```

amaurym commented:
@mergify io backport release/v0.42.x

amaurym commented:
@mergifyio backport release/v0.42.x
